### ISSUE TITLE
Several issues in the tutorial

### ISSUE THREAD

**Issue by chrisfillmore** [google/dagger#1776]
I recently completed the tutorial at https://dagger.dev/tutorial

There are a few continuity gaps which I think make it more likely users will encounter errors. While I acknowledge the learning value of working through those kinds of problems, I don't think the gaps are intentional, and it could lead to users getting confused and not finishing the tutorial. These are the gaps I noticed:

---

Gap 1: CommandRouterFactory abruptly changes from being a top-level interface to being nested on CommandLineAtm
---

In [step 2](https://dagger.dev/tutorial/02-initial-dagger), you are told to create the component interface for CommandRouterFactory
```
@Component interface CommandRouterFactory {
...
```
This is a top-level interface, which is built out by Dagger into `DaggerCommandRouterFactory`, used in CommandLineAtm class:
```
class CommandLineAtm {
  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    CommandRouterFactory commandRouterFactory = DaggerCommandRouterFactory.create();
...
```
CommandRouterFactory is used again in [step 4](https://dagger.dev/tutorial/04-depending-on-interface), when we add modules to the component:
```
@Component(modules = HelloWorldModule.class)
interface CommandRouterFactory {
...
```
But then in [step 5](https://dagger.dev/tutorial/05-abstraction-for-output), `CommandRouterFactory` is suddenly nested on `CommandLineAtm`:
```
class CommandLineAtm {
  ...

  @Component(modules = {HelloWorldModule.class, SystemOutModule.class})
  interface CommandRouterFactory {
...
```
This is important because it changes whether we need to reference `DaggerCommandRouterFactory` or `DaggerCommandLineAtm_CommandRouterFactory`.

---

Gap 2: LoginCommandModule changes from being an abstract class to an interface
---

In [step 6](https://dagger.dev/tutorial/06-new-command), we introduce the abstract class LoginCommandModule:
```
@Module
abstract class LoginCommandModule {
...
```
Later, in [step 14](https://dagger.dev/tutorial/14-avoiding-recursive-logins), `LoginCommandModule` is an interface:
```
@Module
interface LoginCommandModule {
  @BindsOptionalOf
  Account optionalAccount();
}
```
This unexpected change is easy to accommodate, but it's not clear if there's a reason for the change. It seems we could probably just have an interface in the first place.

---

Gap 3: CommandRouterFactory abruptly disappears
---
In [step 10](https://dagger.dev/tutorial/10-deposit-after-login) we introduce the `CommandProcessorFactory`:
```
class CommandLineAtm {
...
  @Component(modules =  ... )
  interface CommandProcessorFactory {
    CommandProcessor processor();
  }
}
```
Because of Gap 1, it's not obvious that we are replacing CommandRouterFactory with CommandProcessorFactory (since CommandRouterFactory was perhaps in its own file). In particular, this part caused a problem for me:
```
@Component(modules =  ... )
interface CommandProcessorFactory {
```
What is `(modules = ...)` supposed to be? I eventually figured this out and was able to move on, but I ran into several errors along the way.

---

There are a variety of other small gaps. These are not such a big deal, but some small notes, or comments in the code indicating that the remaining methods need to be filled in, would help with continuity:
- In [step 9](https://dagger.dev/tutorial/09-maintaining-state) we reference `Account#deposit`, which doesn't exist yet
- Similarly, in [step 10](https://dagger.dev/tutorial/10-deposit-after-login) the definition of `Result` is incomplete, many methods are missing

**Comment by Chang-Eric** [comment]
Thanks for this feedback, this is really helpful. I agree that some of those continuity gaps can be jarring for someone who doesn't know if they are important or not and notes or just making them consistent would be better.

The docs are in the repo but just on the gh-pages branch (once you're in there you should see a tutorial directory). I'd be happy to review a pull request if you have the time to submit one, thanks!

**Comment by cognitiaclaeves** [comment]
Five years later, I'm going through the tutorial and finding it very hard to follow. I seem to be having more trouble than @chrisfillmore , though. My Java knowledge is pretty scattered. I'm on a project that I've been on for about a year now that uses dagger. Prior to that, my exposure to Java was writing Swing and related back-end components, when Java was not a strongly typed language and the TogetherSoft IDE was what generated pattern-boilerplate, instead of a DI library. I find the notion of "invisible patterns" to be very confusing.

My first issues started with incomplete code listings (that didn't include import statements which I had to fish out or just guess at) and lack of context in the git repo. ... I think if the examples were maintained as a separate git repo it may have made things easier. But then, again, when I went to fish for context, I only found a commit that committed everything at once and a commit that recommitted everything after it had been accidentally erased. And one unhelpful commit in between for 'fix build issues' that also seemed to change a mass of files.

It looks as if I won't be able to finish the tutorial, as I've reached a point where I'm unable to figure out what is going on, at step 10. After starting from step 1 (setup) and puzzling my way through to step 10 (deposit after login), there appears to be something happening with Account that I'm unable to figure out what it is:

```
ERROR: /Users/jnorment/projects/bazel/wkspc/dagger/java/dagger/example/atm_jae/BUILD:23:13: Building java/dagger/example/atm_jae/libatm.jar (16 source files) and running annotation processors (ComponentProcessor) failed: (Exit 1): java failed: error executing Javac command (from target //java/dagger/example/atm_jae:atm) external/rules_java~~toolchains~remotejdk21_macos_aarch64/bin/java '--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED' '--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED' ... (remaining 19 arguments skipped)
java/dagger/example/atm_jae/UserCommandsRouter.java:61: error: cannot find symbol
    UserCommandsRouter create(@BindsInstance Account account);
                                             ^
  symbol:   class Account
  location: interface Factory
error: BindsInstanceProcessingStep was unable to process 'account' because 'Account' could not be resolved.
  
  Dependency trace:
      => element (INTERFACE): dagger.example.atm.UserCommandsRouter.Factory
      => element (METHOD): create(Account)
      => element (PARAMETER): account
      => type (ERROR parameter type): Account
  
  If type 'Account' is a generated type, check above for compilation errors that may have prevented the type from being generated. Otherwise, ensure that type 'Account' is on your classpath.
error: ComponentProcessingStep was unable to process 'dagger.example.atm.CommandProcessorFactory' because 'Account' could not be resolved.
  
  Dependency trace:
      => element (INTERFACE): dagger.example.atm.UserCommandsRouter.Factory
      => element (METHOD): create(Account)
      => element (PARAMETER): account
      => type (ERROR parameter type): Account
  
  If type 'Account' is a generated type, check above for compilation errors that may have prevented the type from being generated. Otherwise, ensure that type 'Account' is on your classpath.
error: ComponentProcessingStep was unable to process 'dagger.example.atm.UserCommandsRouter' because 'Account' could not be resolved.
  
  Dependency trace:
      => element (INTERFACE): dagger.example.atm.UserCommandsRouter
      => element (INTERFACE): dagger.example.atm.UserCommandsRouter.Factory
      => element (METHOD): create(Account)
      => element (PARAMETER): account
      => type (ERROR parameter type): Account
  
  If type 'Account' is a generated type, check above for compilation errors that may have prevented the type from being generated. Otherwise, ensure that type 'Account' is on your classpath.
error: ComponentProcessingStep was unable to process 'dagger.example.atm.UserCommandsRouter.Factory' because 'Account' could not be resolved.
  
  Dependency trace:
      => element (INTERFACE): dagger.example.atm.UserCommandsRouter.Factory
      => element (METHOD): create(Account)
      => element (PARAMETER): account
      => type (ERROR parameter type): Account
  
  If type 'Account' is a generated type, check above for compilation errors that may have prevented the type from being generated. Otherwise, ensure that type 'Account' is on your classpath.
java/dagger/example/atm_jae/CommandLineAtm.java:63: error: cannot find symbol
    CommandProcessorFactory commandProcessorFactory = DaggerCommandProcessorFactory.create();
                                                      ^
  symbol:   variable DaggerCommandProcessorFactory
  location: class CommandLineAtm
```

Up until this point, I've been able to figure out that it's usually been an import statement that is missing. Sometimes I can recover this statement, if it still exists in the final version of the code for the example. But sometimes, that line ends up being taken out before the final version of the code, and since there's no history in the git repo of these changes, I'm left to puzzle it out myself. So I guessed at adding this line into `UserCommandsRouter`:

```
import dagger.example.atm.Database.Account;
```

And then got this error:

```
ERROR: /Users/jnorment/projects/bazel/wkspc/dagger/java/dagger/example/atm_jae/BUILD:23:13: Building java/dagger/example/atm_jae/libatm.jar (16 source files) and running annotation processors (ComponentProcessor) failed: (Exit 1): java failed: error executing Javac command (from target //java/dagger/example/atm_jae:atm) external/rules_java~~toolchains~remotejdk21_macos_aarch64/bin/java '--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED' '--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED' ... (remaining 19 arguments skipped)
java/dagger/example/atm_jae/CommandRouter.java:55: error: [Dagger/MissingBinding] dagger.example.atm.Database.Account cannot be provided without an @Inject constructor or an @Provides-annotated method.
interface CommandProcessorFactory {
^
  
      dagger.example.atm.Database.Account is injected at
          [dagger.example.atm.CommandProcessorFactory] dagger.example.atm.DepositCommand(account, …)
      dagger.example.atm.DepositCommand is injected at
          [dagger.example.atm.CommandProcessorFactory] dagger.example.atm.UserCommandsModule.deposit(command)
      java.util.Map<java.lang.String,dagger.example.atm.Command> is injected at
          [dagger.example.atm.CommandProcessorFactory] dagger.example.atm.CommandRouter(commands)
      dagger.example.atm.CommandRouter is injected at
          [dagger.example.atm.CommandProcessorFactory] dagger.example.atm.CommandProcessor(firstCommandRouter)
      dagger.example.atm.CommandProcessor is requested at
          [dagger.example.atm.CommandProcessorFactory] dagger.example.atm.CommandProcessorFactory.commandProcessor()
  The following other entry points also depend on it:
      dagger.example.atm.UserCommandsRouter.router() [dagger.example.atm.CommandProcessorFactory → dagger.example.atm.UserCommandsRouter]
java/dagger/example/atm_jae/CommandLineAtm.java:63: error: cannot find symbol
    CommandProcessorFactory commandProcessorFactory = DaggerCommandProcessorFactory.create();

```

... which seems to be more difficult to solve without a deeper understanding of dagger, which is why I'm here trying to work through the tutorial in the first place. And there's enough of a difference between the tutorial's example and the final source:

tutorial:

```
@Subcomponent(modules = UserCommandsModule.class)
interface UserCommandsRouter {
  CommandRouter router();

  @Subcomponent.Factory
  interface Factory {
    UserCommandsRouter create(@BindsInstance Account account);
  }
```

```
@PerSession
@Subcomponent(modules = {AccountModule.class, AmountsModule.class, UserCommandsModule.class})
interface UserCommandsRouter {
  CommandRouter router();

  @Subcomponent.Factory
  interface Factory {
    UserCommandsRouter create(@BindsInstance @Username String username);
  }
```

Also, there is no mention that Username was its own weird little file in step 10. And there's a lot in that file that I don't understand. It doesn't even appear to be a java code file to me:

```
package dagger.example.atm;

import static java.lang.annotation.RetentionPolicy.RUNTIME;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import javax.inject.Qualifier;

/** Qualifier for the currently logged-in user. */
@Retention(RUNTIME)
@Qualifier
@Documented
@interface Username {}
```

I'm going to try to read through the rest of the tutorial, but because I don't have a working version and do not have any context to reference to get a working a version at this stage, I'm not optimistic that I'll be able to finish the tutorial.

**Comment by cognitiaclaeves** [comment]
The next page, https://dagger.dev/tutorial/13-max-withdrawal-across-commands, mentions `@Scope`. Is that even a thing now? The link for it is broken: `https://dagger.dev/api/latest/dagger/Scope.html` .
